api/pkg/controller/rbac-user-cluster/controller.go

```
@@ -0,0 +1,93 @@
+	"k8s.io/apimachinery/pkg/api/errors"
+)
+
+const (
+	machinedeploymentCRDName = "machinedeployments.cluster.k8s.io"
+	machinesCRDName          = "machines.cluster.k8s.io"
+	controllerName           = "rbac-user-cluster-controller"
+
+	// time in sec
+	requeueAfter1sec = time.Second
+)
+
+// Add creates a new RBAC generator controller that is responsible for creating Role for` machinedeployments` and
+// `machines` resources and RoleBindings for groups: `owners`, `editors` and `viewers``
+func Add(mgr manager.Manager) (string, error) {
+	reconcile := &reconcileRBAC{Client: mgr.GetClient(), ctx: context.TODO()}
+
+	return controllerName, add(mgr, reconcile)
+}
+
+// add adds a new Controller to mgr with r as the reconcile.reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to CRDs
+	err = c.Watch(&source.Kind{Type: &apiextensionv1beta1.CustomResourceDefinition{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// reconcileRBAC reconciles all RBAC related objects
+type reconcileRBAC struct {
+	ctx context.Context
+	client.Client
+}
+
+// Reconcile makes changes in response to RBAC related object changes
+func (r *reconcileRBAC) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	var err error
+
+	rdr := reconciler{client: r.Client, ctx: r.ctx}
+	crd := &apiextensionv1beta1.CustomResourceDefinition{}
+	err = r.Get(r.ctx, request.NamespacedName, crd)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{
+			//  there is no need to set Requeue to true at the same time as RequeueAfter
+			// TODO: figure out how to requeue in exponential manner
+			RequeueAfter: requeueAfter1sec,
+		}, nil
```

I _believe_ the only thing that has to be done here is to actually return the error instead of swallowing it, the controller will then call `Reconcile` again: https://github.com/kubernetes-sigs/controller-runtime/blob/bd4677326d3a1e201f986bf798638fd38fbe3680/pkg/reconcile/reconcile.go#L82

CC @nikhita because I believe this has to be fixed in the user cluster controller as well

---

good point, thanks

---

Yeah it just uses the `AddRateLimited` from the workqueue if there was an error: https://github.com/kubernetes-sigs/controller-runtime/blob/0dd6edb02bf23e3669a3edc50ef00a38719173b0/pkg/internal/controller/controller.go#L216

---

Re user cluster controller, it returns an error:

https://github.com/kubermatic/kubermatic/blob/3da7568e8aa898eb836f5182611e8dbc115b56d3/api/pkg/controller/usercluster/controller.go#L54-L58

---

Maybe just remove the `RequeueAfter` then? It is a no-op right now but would make the code easier to understand if it wasn't there

---

we decided to use `RequeueAfter: requeueAfter1sec` in some unexpected cases like a network problem to delay another check

---

I would argue every error that happens here is unexpected and the normal backoff mechanism of the q should be used instead of trying to be smarter

---

ok, so I will remove it